File: DentalWave-frontend/src/services/UserService.js
import axios from 'axios'
import { getAuthHeader } from './AuthService'

/**
 * Base URL for User endpoints.
 */
const USER_REST_API_BASE_URL = 'http://localhost:8080/api/users'

/**
 * Creates a new user.
 *
 * @param user UserDto
 * @returns Created UserDto
 */
export const createUser = (user) =>
    axios.post(USER_REST_API_BASE_URL, user, getAuthHeader())

/**
 * Retrieves a user by id.
 *
 * @param id User id
 * @returns UserDto
 */
export const getUserById = (id) =>
    axios.get(`${USER_REST_API_BASE_URL}/${id}`, getAuthHeader())

/**
 * Retrieves the currently logged-in user's profile.
 *
 * @returns UserDto
 */
export const getCurrentUser = () =>
    axios.get(`${USER_REST_API_BASE_URL}/me`, getAuthHeader())

/**
 * Retrieves all users.
 *
 * @returns List<UserDto>
 */
export const getAllUsers = () =>
    axios.get(USER_REST_API_BASE_URL, getAuthHeader())

/**
 * Updates a user.
 *
 * @param id User id
 * @param user Updated UserDto
 * @returns Updated UserDto
 */
export const updateUser = (id, user) =>
    axios.put(`${USER_REST_API_BASE_URL}/${id}`, user, getAuthHeader())

/**
 * Updates the currently logged-in user's editable profile fields.
 *
 * @param user Updated UserDto fields
 * @returns Updated UserDto
 */
export const updateCurrentUser = (user) =>
    axios.put(`${USER_REST_API_BASE_URL}/me`, user, getAuthHeader())

/**
 * Deletes a user.
 *
 * @param id User id
 */
export const deleteUser = (id) =>
    axios.delete(`${USER_REST_API_BASE_URL}/${id}`, getAuthHeader())

/**
 * Retrieves a user by email.
 *
 * @param email User email
 * @returns UserDto
 */
export const getUserByEmail = (email) =>
    axios.get(`${USER_REST_API_BASE_URL}/email/${email}`, getAuthHeader())

/**
 * Retrieves users by role.
 *
 * @param role Role name
 * @returns List<UserDto>
 */
export const getUsersByRole = (role) =>
    axios.get(`${USER_REST_API_BASE_URL}/role/${role}`, getAuthHeader())

/**
 * Searches users by keyword.
 *
 * Example:
 * searchUsers("smith")
 *
 * @param keyword Search term
 * @returns List<UserDto>
 */
export const searchUsers = (keyword) =>
    axios.get(`${USER_REST_API_BASE_URL}/search?keyword=${keyword}`, getAuthHeader())

/**
 * Enables a user account.
 *
 * @param id User id
 * @returns Updated UserDto
 */
export const enableUser = (id) =>
    axios.put(`${USER_REST_API_BASE_URL}/${id}/enable`, null, getAuthHeader())

/**
 * Disables a user account.
 *
 * @param id User id
 * @returns Updated UserDto
 */
export const disableUser = (id) =>
    axios.put(`${USER_REST_API_BASE_URL}/${id}/disable`, null, getAuthHeader())

File: DentalWave/src/main/java/com/summerproject2026/DentalWave/service/UserService.java
package com.summerproject2026.DentalWave.service;

import java.util.List;
import com.summerproject2026.DentalWave.dto.UserDto;

/**
 * Service interface for managing users within the DentalWave system.
 *
 * Responsibilities:
 * <ul>
 *   <li>Create, retrieve, update, and delete users.</li>
 *   <li>Search for users by username, role, or keyword.</li>
 *   <li>Enable and disable user accounts.</li>
 * </ul>
 */
public interface UserService {

    /**
     * Creates a new user.
     *
     * @param userDto the user information to create
     * @return the newly created user
     */
    UserDto createUser(UserDto userDto);

    /**
     * Retrieves a user by their unique identifier.
     *
     * @param id the user id
     * @return the matching user
     */
    UserDto getUserById(Long id);

    /**
     * Retrieves all users in the system.
     *
     * @return a list of all users
     */
    List<UserDto> getAllUsers();

    /**
     * Updates an existing user's information.
     *
     * @param id the id of the user to update
     * @param userDto the updated user information
     * @return the updated user
     */
    UserDto updateUser(Long id, UserDto userDto);

    /**
     * Deletes a user from the system.
     *
     * @param id the id of the user to delete
     */
    void deleteUser(Long id);

    /**
     * Retrieves a user by their username.
     *
     * @param username the username to search for
     * @return the matching user
     */
    UserDto getUserByUsername(String username);

    /**
     * Retrieves a user by their email address.
     *
     * @param email the email address to search for
     * @return the matching user
     */
    UserDto getUserByEmail(String email);

    /**
     * Retrieves all users assigned to a specific role.
     *
     * @param role the role name
     * @return a list of users with the specified role
     */
    List<UserDto> getUsersByRole(String role);

    /**
     * Searches for users matching the provided keyword.
     *
     * @param keyword the search term
     * @return a list of matching users
     */
    List<UserDto> searchUsers(String keyword);

    /**
     * Enables a user account.
     *
     * @param id the id of the user to enable
     * @return the updated user with enabled status set to true
     */
    UserDto enableUser(Long id);

    /**
     * Disables a user account.
     *
     * @param id the id of the user to disable
     * @return the updated user with enabled status set to false
     */
    UserDto disableUser(Long id);
}